src/event_loop/macos.rs

```diff
@@ -0,0 +1,137 @@
+//! An event loop implementation for macOS.
+
+use core_foundation::base::kCFAllocatorDefault;
+use core_foundation::runloop::{
+    kCFRunLoopCommonModes, CFRunLoopAddSource, CFRunLoopGetMain, CFRunLoopRemoveSource,
+    CFRunLoopSourceContext, CFRunLoopSourceCreate, CFRunLoopSourceInvalidate, CFRunLoopSourceRef,
+    CFRunLoopSourceSignal, CFRunLoopWakeUp,
+};
+use crossbeam::channel::{self, Receiver, Sender};
+use objc::{class, msg_send, sel, sel_impl};
+use std::os::raw::c_void;
+use std::sync::Arc;
+
+use super::{BackgroundThread, EventLoop, MainThreadExecutor};
+
+/// Wrapping the `CFRunLoopSourceRef` type is required to be able to annotate it as thread-safe.
+struct LoopSourceWrapper(CFRunLoopSourceRef);
+
+unsafe impl Send for LoopSourceWrapper {}
+unsafe impl Sync for LoopSourceWrapper {}
+
+/// See [`EventLoop`][super::EventLoop].
+pub(crate) struct MacOSEventLoop<T, E> {
+    /// The thing that ends up executing these tasks. The tasks are usually executed from the worker
+    /// thread, but if the current thread is the main thread then the task cna also be executed
+    /// directly.
+    executor: Arc<E>,
+
+    /// A background thread for running tasks independently from the host's GUI thread. Useful for
+    /// longer, blocking tasks.
+    background_thread: BackgroundThread<T>,
+
+    /// The reference to the run-loop source so that it can be torn down when this struct is dropped.
+    loop_source: LoopSourceWrapper,
+
+    /// The sender for passing messages to the main thread.
+    main_thread_sender: Sender<T>,
+
+    /// Is it necessary to store the information that is sent to the callback function here?
+    _info: Arc<(Arc<E>, Receiver<T>)>,
```

Comment:
This can be a `Box` and unless you're deep into Apple's terminology `_info` doesn't really tell you why it's there. But I guess that's just style and I'll change that after merging.

Reply:
i ended up using `Arc` here since `Arc::as_ptr()` doesn't consume the pointer unlike `Box::into_raw()`. is there another clean way to cast a Box into a `*mut c_void` without consuming?

good point about the variable name being quite vague!

Reply:
Borrow the data as an immutable reference. References silently decay into their equivalent raw pointers, but to get to `*mut c_void` you need to explicitly cast it. `&*boxed_data as *const _ as *mut c_void` should do it.